File: ruby/lib/pay_core/rfc3339_parser.rb
# frozen_string_literal: true

require "time"
require "date"

module PayCore
  # RFC 3339 date-time parser shared by solana-mpp and solana-x402.
  #
  # @see https://datatracker.ietf.org/doc/html/rfc3339 RFC 3339 Date and Time on the Internet
  module Rfc3339Parser
    # Strict RFC 3339 date-time (sec 5.6). Year is exactly 4 digits; T literal
    # accepted upper or lower (per parse SHOULD); time-secfrac is "." 1*DIGIT,
    # so the digit count is unbounded here and clamped after the match.
    REGEX = /\A
      (\d{4})-(\d{2})-(\d{2})         # full-date
      [Tt]
      (\d{2}):(\d{2}):(\d{2})         # partial-time
      (?:\.(\d+))?                    # time-secfrac
      (?:[Zz]|([+-]\d{2}):(\d{2}))    # time-offset
      \z/x
    private_constant :REGEX

    module_function

    # Parse an RFC 3339 timestamp into a Time, or nil when the input is
    # not a valid RFC 3339 date-time. Returns nil for any out-of-range
    # component so callers can fail-closed.
    def parse(value)
      return nil unless value.is_a?(String)

      match = REGEX.match(value)
      return nil unless match

      year, month, day = match[1].to_i, match[2].to_i, match[3].to_i
      hour, minute, second = match[4].to_i, match[5].to_i, match[6].to_i
      return nil if month < 1 || month > 12
      return nil if day < 1 || day > 31
      return nil if hour > 23 || minute > 59 || second > 60
      # time-numoffset bounds the offset at 23:59, but Time.iso8601 reads
      # +00:60 as +01:00 — a silent hour shift in the instant an expiry is
      # compared against, so the offset is ranged before it is delegated.
      return nil if match[8] && (match[8].delete("+-").to_i > 23 || match[9].to_i > 59)
      return nil if year > 9999
      return nil unless Date.valid_date?(year, month, day)

      # Rebuilt from the captures rather than delegated verbatim: Time.iso8601
      # rejects the lowercase t/z that sec 5.6 allows, keeps sub-nanosecond
      # secfrac as a Rational no other SDK can carry, and rolls seconds = 60
      # forward into the next minute instead of clamping it.
      leap_second = second == 60
      secfrac = match[7] ? ".#{match[7][0, 9]}" : ""
      secfrac = ".999999999" if leap_second
      offset = match[8] ? "#{match[8]}:#{match[9]}" : "Z"
      parsed = Time.iso8601("#{match[1]}-#{match[2]}-#{match[3]}T#{match[4]}:#{match[5]}:" \
                            "#{leap_second ? "59" : match[6]}#{secfrac}#{offset}")
      return parsed unless leap_second

      # RFC 3339 sec 5.7: a leap second only ever ends a UTC month, and the
      # offset is applied before that is judged. Matches the TypeScript SDK.
      utc = parsed.getutc
      return nil unless utc.hour == 23 && utc.min == 59 && utc.day == Date.new(utc.year, utc.month, -1).day

      parsed
    rescue ArgumentError
      nil
    end
  end
end

File: php/src/PayCore/Rfc3339Parser.php
<?php

declare(strict_types=1);

namespace PayKit\PayCore;

use DateTimeImmutable;
use DateTimeZone;

/**
 * RFC 3339 date-time parser used by Challenge::isExpired().
 *
 * Extracted from Challenge.php per PR #102 review (inline comment 3298110199)
 * so RFC parsing logic lives in a dedicated file. Lua already keeps the
 * parser in lua/mpp/expires.lua; Ruby moves the regex to Rfc3339Parser
 * in the same review round.
 *
 * @see https://datatracker.ietf.org/doc/html/rfc3339 RFC 3339 Date and Time on the Internet
 */
final class Rfc3339Parser
{
    /**
     * Strict RFC 3339 date-time grammar (sec 5.6); accepts lowercase t/z on
     * parse (RFC 3339 sec 5.6 PARSE caveat), year exactly 4 digits,
     * fractional seconds 1 or more digits (time-secfrac = 1*DIGIT).
     */
    private const RFC3339_PATTERN = '/^(\d{4})-(\d{2})-(\d{2})[Tt](\d{2}):(\d{2}):(\d{2})(?:\.(\d+))?(Z|z|([+-])(\d{2}):(\d{2}))$/';

    private function __construct()
    {
    }

    /**
     * Parse an RFC 3339 timestamp into a DateTimeImmutable, or null when the
     * input is not a valid RFC 3339 date-time. Returns null for any
     * out-of-range component so callers can fail-closed.
     */
    public static function parse(string $value): ?DateTimeImmutable
    {
        if (preg_match(self::RFC3339_PATTERN, $value, $m) !== 1) {
            return null;
        }
        $year = (int)$m[1];
        $month = (int)$m[2];
        $day = (int)$m[3];
        $hour = (int)$m[4];
        $minute = (int)$m[5];
        $second = (int)$m[6];
        $offsetTag = $m[8];
        // RFC 3339 section 5.7 allows seconds = 60 for positive leap seconds.
        // Lua and Go SDKs accept the value at the parser level; PHP must too,
        // otherwise a credential timestamped exactly at 23:59:60 UTC is
        // wrongly flagged expired.
        if ($month < 1 || $month > 12 || $day < 1 || $day > 31 || $hour > 23 || $minute > 59 || $second > 60) {
            return null;
        }
        // year 0000 is a leap year (date-fullyear = 4DIGIT); checkdate() rejects it, so validate against leap-year 4.
        if ($year > 9999 || !checkdate($month, $day, $year === 0 ? 4 : $year)) {
            return null;
        }
        if ($offsetTag !== 'Z' && $offsetTag !== 'z') {
            $offHour = isset($m[10]) ? (int)$m[10] : 0;
            $offMin = isset($m[11]) ? (int)$m[11] : 0;
            if ($offHour > 23 || $offMin > 59) {
                return null;
            }
        }

        // Normalize lowercase t/z to uppercase before delegating to DateTimeImmutable (DATE_ATOM is strict).
        // PHP's `u` format only accepts exactly six fractional digits; RFC 3339 permits 1..9.
        // Truncate the regex-captured fractional component to microseconds (the regex already
        // bounded it to 1..9 digits, so truncation is safe). Sub-microsecond precision is dropped
        // for expiry comparison purposes which is acceptable since we only need second-level resolution.
        $normalized = strtr($value, ['t' => 'T', 'z' => 'Z']);
        // RFC 3339 §5.7 leap second normalization. The range guard above
        // accepts sec=60 for spec parity with Lua/Go/Ruby, but PHP's
        // DateTimeImmutable::createFromFormat rejects :60 outright (it
        // does not implement the leap-second extension). Downshift to
        // :59 before delegating so a credential timestamped exactly at
        // 23:59:60 UTC parses to 23:59:59 UTC for expiry comparison.
        // The 1-second resolution slip at the leap-second boundary has
        // no operational impact: expiry comparison is whole-second.
        if ($second === 60) {
            $normalized = preg_replace('/:60(?=\.|Z|[+\-])/', ':59', $normalized, 1) ?? $normalized;
        }
        $frac = $m[7];
        if ($frac !== '') {
            $truncated = substr($frac, 0, 6);
            $normalized = preg_replace('/\.\d+/', '.' . $truncated, $normalized, 1) ?? $normalized;
        }
        $parsed = DateTimeImmutable::createFromFormat(DATE_ATOM, $normalized);
        if ($parsed === false) {
            $parsed = DateTimeImmutable::createFromFormat('Y-m-d\TH:i:s.up', $normalized);
        }
        if ($parsed === false) {
            $parsed = DateTimeImmutable::createFromFormat('Y-m-d\TH:i:s.uP', $normalized);
        }
        if ($parsed === false) {
            return null;
        }
        if ($second === 60) {
            $utc = $parsed->setTimezone(new DateTimeZone('UTC'));
            if ($utc->format('H') !== '23' || $utc->format('i') !== '59' || $utc->format('j') !== $utc->format('t')) {
                return null;
            }
        }
        return $parsed;
    }
}
